waiter/src/waiter/service_description.clj

```diff
@@ -1222,21 +1223,25 @@
       (> (+ auth-timestamp (-> consent-expiry-days t/days t/in-millis))
          (.getMillis ^DateTime (clock))))))
 
-(defn source-tokens->idle-timeout
-  "Computes the idle timeout of the service using token parameters and defaults."
+(defn- source-tokens->gc-time
+  "Computes the time when a service created using provided source tokens should be GC-ed.
+   It ignores whether GC has been disabled by configuring idle-timeout-mins=0."
   [token->token-parameters attach-token-defaults-fn source-tokens-seq]
   (->> source-tokens-seq
-    (map (fn source-tokens->idle-timeout [source-tokens]
+    (map (fn source-tokens->gc-time-helper [source-tokens]
            (let [{:strs [fallback-period-secs stale-timeout-mins]}
                  (->> source-tokens
                    (map #(some-> % :token token->token-parameters))
                    (reduce merge {})
-                   (attach-token-defaults-fn))]
-             (+ (-> (+ fallback-period-secs (dec (-> 1 t/minutes t/in-seconds))) ;; ceiling
-                  t/seconds
-                  t/in-minutes)
-                stale-timeout-mins))))
-    (reduce max)))
+                   (attach-token-defaults-fn))
+                 most-recent-token-update-time
+                 (->> source-tokens
+                   (map #(some-> % :token token->token-parameters (get "last-update-time")))
+                   (reduce max))]
```

Comment:
Should we use `keep` instead of `map` here? The `(reduce max)` is going to NPE if there are `nil`s in the sequence.

Reply:
Changed to use `keep` and check for empty sequence.
This can happen when the token has been deleted.
Also added unit test.